fetch.bs

```diff
@@ -1567,13 +1573,13 @@ whose <a for=request>destination</a> is "<code>audio</code>", "<code>audioworkle
 "<code>object</code>" or "<code>embed</code>".
 
 <p>A <dfn export>non-subresource request</dfn> is a <a for=/>request</a>
-whose <a for=request>destination</a> is "<code>document</code>",
-"<code>report</code>", "<code>serviceworker</code>", "<code>sharedworker</code>",
-or "<code>worker</code>".
+whose <a for=request>destination</a> is "<code>document</code>", "<code>frame</code>",
+"<code>iframe</code>", "<code>report</code>", "<code>serviceworker</code>",
+"<code>sharedworker</code>", or "<code>worker</code>".
```

Comment:
This isn't accurate anymore. Currently for embed/object the destination would be document if they have a browsing context that is navigated and they would end up classified here as a non-subresource request.

Reply:
Currently, `embed` and `object` are "potential-navigation-or-subresource" requests, and don't get categorized into either "subresource" or "non-subresource". I don't think this patch changes that.

Looking at "display a plugin", it looks like `<embed>` and `<object>` will always create a nested browsing context. Is that how we're categorizing "subresource" vs "non-subresource"? If so, then I agree that it seems like we should lump them into the "non-subresource" bucket.

It's not clear to me what the impact is there, though. Where is "non-subresource request" used?

Reply:
My understanding is that `embed` (and `object`) are initially no-cors / embed, but change to navigate / document the moment they display a non-XXX resource (maybe XXX is only images?) and instantiate a nested browsing context.

https://w3c.github.io/ServiceWorker/ uses non-subresource.

Reply:
As far as Fetch is concerned, I think `<embed>` and `<object>` are both always `no-cors` / {`embed`,`object`}.

Tracing through from https://html.spec.whatwg.org/#the-embed-element-setup-steps:

1.  Step 3.4 fetches with `no-cors` / `embed`.
2.  Step 3.4.3 switches on the type of the response (in a way that I'm not sure matches implementations; both Firefox and Chrome render `text/html`, for instance) to either [display a plugin](https://html.spec.whatwg.org/#display-a-plugin) or navigate to the response.
3.  [Displaying a plugin](https://html.spec.whatwg.org/#display-a-plugin) drops any nested browsing context in step 1, and hand-waves at finding a plugin for the response type.
4.  Navigating to the response does not re-fetch with `navigation` / `document`; it navigates to the response that was given for the `no-cors`/`embed` fetch above.

`<object>` looks pretty similar, though more capacious with regard to the types it handles specially.

(It's probably worth reevaluating all of this complexity in the light of Flash's impending death. Perhaps we can make `<object>` and `<embed>` act like strangely-spelled `<iframe>`s at some point?)

> https://w3c.github.io/ServiceWorker/ uses non-subresource.

https://w3c.github.io/ServiceWorker/#on-fetch-request-algorithm uses "non-subresource request" only after punting entirely on "potential-navigation-or-subresource" requests, which means that the categorization would have no practical effect there.

Reply:
If you navigate a browsing context whose owner is `embed`, how does it end up with `no-cors` / `embed`?

Reply:
Huh. I hadn't thought about that case. My initial thoughts are:

1. Don't do that.
2. That's surprising behavior!

Should we special-case this in the navigate algorithm by checking `<embed>` and `<object>` in https://github.com/whatwg/html/pull/4976 along with `<iframe>` and `<frame>`?

Reply:
I think it makes sense that it becomes `navigate` as at that point it is navigation and cannot be used as a subresource anymore. It might be okay that it stays `embed`/`object` though if we go that way for `frame`/`iframe`.

Reply:
So, given that the potential resolution to the underlying question in https://github.com/whatwg/fetch/pull/948#issuecomment-543560997 puts the `embed` and `object` destinations into two categories (depending on the mode), what would you like me to do with these definitions ("navigation request", "non-subresource request"?

Should we leave them as they are in this patch, or do you want to include/excude requests with a `mode` of `navigate` as appropriate?

Reply:
Coming back to this: since we've decided that `mode` for `embed` and `object` is going to be `navigate`, do you still want changes in this section, or are you comfortable with the current text of this PR?

Reply:
Ok. Is that something you want me to do in _this_ patch, which introduces `iframe` and `frame`? Or would you be ok with me splitting it out into a separate patch (which I think I'd like to do, because some things about changing `<embed>`'s mode to `navigate` aren't clear to me, and will likely require changes in both Fetch and HTML).

If you're alright with deferring it to a future patch, is there anything else we need to change about this one?

Reply:
I'm a little worried it's gonna linger and never get fixed, despite being somewhat important from an interoperability perspective.

And note that before it'd go from `no-cors`/`embed` to `navigate`/`document` and that is being changed too.

And given that it also changes for `iframe` et al, there needs to be an associated HTML PR already for this change.

Reply:
> I'm a little worried it's gonna linger and never get fixed, despite being somewhat important from an interoperability perspective.
>
> And note that before it'd go from no-cors/embed to navigate/document and that is being changed too.

I agree that we should fix it in the spec and lock it in with tests. I'm not sure _this_ patch is the right place to do it, given how simple it is today, and how not-simple I think the `<embed>` change is going to be more generally. :)

If you'd prefer one big patch that addresses `<embed>`/`<object>`/`<frame>`/`<iframe>` all together, I'll do that! But I think I'd prefer to just land this and the associated HTML PR, and worry about `<embed>` and `<object>` in a subsequent set of PRs.

> And given that it also changes for iframe et al, there needs to be an associated HTML PR already for this change.

https://github.com/whatwg/html/pull/4976 is that change.

Reply:
I might be missing something, but apart from the tests I don't think it's as big of a change on this side? The HTML side would require a lot of work, but not sure all of that is blocking given the many issues embed/object have already.

Reply:
Step 5 of main fetch (https://fetch.spec.whatwg.org/#main-fetch) has pretty different behavior between "navigate" and "no-cors". I think it'll turn out to be incorrect for the initial request from `<embed>`, but I haven't through through what needs to change (or whether I'm even correct that it does need to change).

Reply:
(There's also a minor MIME type change in step 1.3.2 of Fetching (https://fetch.spec.whatwg.org/#fetching), but that's likely not a big deal to change.)

Reply:
Ah, what `Accept` should say.